components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -336,189 +604,254 @@ impl RangeManager {
-            info!(
-                "evict a range that is not cached";
-                "range" => ?evict_range,
-            );
-            return vec![];
+        let mut deleteable_regions = vec![];
+        let mut evict_ids = vec![];
+        let evict_range = CacheRange::from_region(evict_region);
+        self.iter_overlapped_regions(&evict_range, |meta| {
+            if evict_range.start >= meta.range.end {
+                return false;
+            }
+            evict_ids.push(meta.region.id);
+            true
+        });
+        if evict_ids.is_empty() {
+            info!("evict a region that is not cached"; 
+                "reason" => ?evict_reason, 
+                "region" => ?evict_region);
         }
-
-        overlapped_ranges
-            .into_iter()
-            .filter(|r| self.evict_within_range(r, r, evict_reason))
-            .collect()
+        for rid in evict_ids {
+            if let Some(region) = self.do_evict_region(rid, evict_region, evict_reason) {
+                deleteable_regions.push(region);
+            }
+        }
+        deleteable_regions
     }
 
-    // Return true means there is no ongoing snapshot, the evicted_range can be
-    // deleted now.
-    fn evict_within_range(
+    // return the region if it can be directly deleted.
+    fn do_evict_region(
         &mut self,
-        evict_range: &CacheRange,
-        cached_range: &CacheRange,
+        id: u64,
+        evict_region: &Region,
         evict_reason: EvictReason,
-    ) -> bool {
-        assert!(cached_range.contains_range(evict_range));
-        info!(
-            "evict range in cache range engine";
-            "evict_range" => ?evict_range,
-            "cached_range" => ?cached_range,
+    ) -> Option<Region> {
+        let meta = self.regions.get_mut(&id).unwrap();
+        let prev_state = meta.state;
+        assert!(
+            meta.range.overlaps(&CacheRange::from_region(evict_region)),
+            "meta: {:?}, evict_region: {:?}",
+            meta,
+            evict_region
         );
-        let meta = self.ranges.remove(cached_range).unwrap();
-        let (left_range, right_range) = cached_range.split_off(evict_range);
-        assert!((left_range.is_some() || right_range.is_some()) || evict_range == cached_range);
-
-        if let Some(left_range) = left_range {
-            let left_meta = RangeMeta::derive_from(self.id_allocator.allocate_id(), &meta);
-            self.ranges.insert(left_range, left_meta);
+        if prev_state == RegionState::Pending {
+            let meta = self.remove_region(id);
+            info!(
+                "evict overlap pending region in cache range engine";
+                "reason" => ?evict_reason,
+                "target_region" => ?evict_region,
+                "overlap_region" => ?meta.region,
+                "state" => ?prev_state,
+            );
+            return None;
+        } else if prev_state >= RegionState::LoadingCanceled {
```

Comment:
I prefer not using Ordering between state. Reader needs to remember the order of them and if we are to add a new state in the future, it's easy to make mistakes.

Reply:
Done.